### PR TITLE
Ft rectangles: still one issue about line_to_line!

### PR DESCRIPTION
**Pull request by saraybakker1**
Added rectangle distances to capsule, points, spheres.
wondering if function Line_to_line (already in fabrics) is correct??? (I am using this function in the capsule_to_rectangle computation). Added tests for the functions
If bug is fixed, these tests (ow in test_distances_rectangular.py) can be added to the test_distances.py file.

### CODE REVIEW

**Comment by saraybakker1** [comment]
Comment about why I think the line_to_line doesn't work in 3d space. If you have two lines in 2D space that do not intersect, then the closest point will be one of the edge points. However, in 3d space, if a line is orthogonal (z-plane) while another line is in the (xy)-plane, it can happen that the closest point of a line segment is somewhere halfway the line. (see the test case of the rectangle and capsule)

File: fabrics/helpers/distances.py [hunk 1]
```diff
@@ -88,3 +88,78 @@ def capsule_to_sphere(
...
+
+def capsule_to_rectangle(
+        capsule_centers: List[ca.SX],
+        rect: dict,
+        capsule_radius: ca.SX
+):
+    min_distance = ca.inf
+    for corner_line in rect["corner_lines"]:
+        dist = line_to_line(corner_line[0], corner_line[1], capsule_centers[0], capsule_centers[1]) - capsule_radius
+        min_distance = min(dist, min_distance)
```
Comment (maxspahn) [inline fabrics/helpers/distances.py:164]: There is a wrong assumption here: The distance between a cuboid and a line is not the minimum of the distances to all edges, as the distance to the face plane might be shorter

**Review by maxspahn — CHANGES_REQUESTED** [review]
I will fix the types and the wrong function to compute the distance between capsule and cuboid.
No need for action from your side.

**Comment by maxspahn** [comment]
Very nice that you noticed the error in `line_to_line`. There is hardly a good solution to it. For now, I just rely on sampling points on one of the lines.

**Comment by maxspahn** [comment]
@saraybakker1 I made some major changes to the tests and the function itself to be consistent when to use casadi and when to use numpy. 

Would be willing to make an example for the rectangle collision avoidance?

**Comment by xueminchi** [comment]
> Hi @maxspahn , I can make a nice example. However, will do that later this week :) I can also look at some line-line distances that doesn't require samples later :)

Hi @saraybakker1 , I'm very familiar with the distance between point vs polytopes, and potypoes vs polytopes, which includes hyperplane-to-hyperplane, or segment-to-segment. If you are interested, I can send you some materials to your email.

**Comment by maxspahn** [comment]
@xueminchi Can you just post the link to the materials here?

**Comment by xueminchi** [comment]
@maxspahn sure. 
Let's consider 2D cases, and 3D cases. In 2D, signed distance of segment-to-segment can be found here: [link1](https://www.youtube.com/watch?v=PMltMdi1Wzg), other geometries can be found in [link2](https://iquilezles.org/articles/distfunctions2d/). For 3D, if it's point-to-polytope or polytope-to-polytope, I suggest GJK and EPA algorithm, related recent paper [link](https://hal.science/hal-04070039/document), if it's more complex shapes, SDF is needed in [link](https://iquilezles.org/articles/distfunctions/).

**Comment by maxspahn** [comment]
Thanks @xueminchi ,

> Let's consider 2D cases, and 3D cases. In 2D, signed distance of segment-to-segment can be found here: [link1](https://www.youtube.com/watch?v=PMltMdi1Wzg), other geometries can be found in [link2](https://iquilezles.org/articles/distfunctions2d/). F

This only covers the case point-to-line as the generation of the sdf of a capsule is effectively computing the distance line to point for a grid of points.

We are looking for something line-to-line to compute the distance between a capsule and a rectangle.

**Comment by xueminchi** [comment]
@maxspahn Oh, I see. For segment-to-segment, please refer to book <<Real-Time Collision Detection>> author Christer Ericson, page 148. Line-to-Line is simpler, page 146.

**Comment by maxspahn** [comment]
```pseudo 
// Clamp n to lie within the range [min, max]
float Clamp(float n, float min, float max) {
 if (n < min) return min;
 if (n > max) return max;
 return n;
}
 // Computes closest points C1 and C2 of S1(s)=P1+s*(Q1-P1) and
 // S2(t)=P2+t*(Q2-P2), returning s and t. Function result is squared
 // distance between between S1(s) and S2(t)
float ClosestPtSegmentSegment(Point p1, Point q1, Point p2, Point q2,
 float &s, float &t, Point &c1, Point &c2)
{
 Vector d1 = q1 - p1; // Direction vector of segment S1
 Vector d2 = q2 - p2; // Direction vector of segment S2
 Vector r = p1 - p2;
 float a = Dot(d1, d1); // Squared length of segment S1, always nonnegative
 float e = Dot(d2, d2); // Squared length of segment S2, always nonnegative
 float f = Dot(d2, r);
 // Check if either or both segments degenerate into points
 if (a <= EPSILON && e <= EPSILON) {
 // Both segments degenerate into points
  s=t= 0.0f;
  c1 = p1;
  c2 = p2;
  return Dot(c1 - c2, c1 - c2);
 }
 if (a <= EPSILON) {
 // First segment degenerates into a point
  s = 0.0f;
  t = f / e; // s = 0 => t = (b*s + f) / e = f / e
  t = Clamp(t, 0.0f, 1.0f);
 } else {
  float c = Dot(d1, r);
  if (e <= EPSILON) {
 // Second segment degenerates into a point
  t = 0.0f;
  s = Clamp(-c / a, 0.0f, 1.0f); // t = 0 => s = (b*t - c) / a = -c / a
 } else {
// The general nondegenerate case starts here
float b = Dot(d1, d2);
float denom = a*e-b*b; // Always nonnegative
// If segments not parallel, compute closest point on L1 to L2 and
// clamp to segment S1. Else pick arbitrary s (here 0)
if (denom != 0.0f) {
s = Clamp((b*f - c*e) / denom, 0.0f, 1.0f);
} else s = 0.0f;
// Compute point on L2 closest to S1(s) using
// t = Dot((P1 + D1*s) - P2,D2) / Dot(D2,D2) = (b*s + f) / e
t = (b*s + f) / e;
// If t in [0,1] done. Else clamp t, recompute s for the new value
// of t using s = Dot((P2 + D2*t) - P1,D1) / Dot(D1,D1)= (t*b - c) / a
// and clamp s to [0, 1]
if (t < 0.0f) {
t = 0.0f;
s = Clamp(-c / a, 0.0f, 1.0f);
} else if (t > 1.0f) {
t = 1.0f;
s = Clamp((b - c) / a, 0.0f, 1.0f);
}
}
}
c1 = p1 + d1 * s;
c2 = p2 + d2 * t;
return Dot(c1 - c2, c1 - c2);
}
```
from the book

**Comment by maxspahn** [comment]
@xueminchi The distance from the book works. Now, we need something for line to cuboid. You also have some ideas on that?

File: fabrics/helpers/distances.py [hunk 2]
```diff
@@ -87,4 +173,136 @@ def capsule_to_sphere(
...
+def cuboid_to_line(
+    cuboid_center: ca.SX,
+    cuboid_size: ca.SX,
+    line_start: ca.SX,
+    line_end: ca.SX,
+) -> ca.SX:
+    distance = ca.fmin(
+        cuboid_to_point(cuboid_center, cuboid_size, line_start),
+        cuboid_to_point(cuboid_center, cuboid_size, line_end),
+    )
```
Comment (maxspahn) [inline fabrics/helpers/distances.py:271]: @saraybakker1 That was the only thing missing in your formula. It covers the case that the end of the line segment is in front of one of the side faces. 
I copied your initial implementation for the edge-to-line distance to cover the other cases.

**Pull request merged**